File: frontend/src/types/operational-analytics.ts
/**
 * Frontend-owned types for the `/api/v1/analytics/operational/*`
 * resources (Milestone 12) — defined independently of the backend's
 * Pydantic schemas, matching the convention `types/analytics.ts`
 * already follows.
 */

import type { TimeSeries } from "@/types/analytics";
import type { UserMetrics } from "@/types/analytics";
import type { UserRole } from "@/types/workflow";

export interface PendingApprovalAge {
  stage_id: string;
  request_id: string;
  request_title: string;
  request_type: string;
  department: string | null;
  stage_name: string;
  stage_order: number;
  assigned_to: string | null;
  assigned_role: UserRole | null;
  created_at: string;
  age_hours: number;
  sla_hours: number | null;
  is_overdue: boolean;
}

export interface SLAMetrics {
  sla_hours_override: number | null;
  pending_stage_count: number;
  overdue_stage_count: number;
  overdue_request_count: number;
  average_current_stage_age_hours: number | null;
  decided_stage_count: number;
  sla_breaches_decided: number;
  sla_compliance_percentage: number | null;
  average_total_workflow_duration_seconds: number | null;
  average_approval_duration_seconds: number | null;
}

export interface DurationBucket {
  key: string;
  average_seconds: number | null;
  count: number;
}

export interface ApprovalDelayReport {
  longest_pending: PendingApprovalAge[];
  oldest_pending_requests: PendingApprovalAge[];
  average_approval_seconds: number | null;
  median_approval_seconds: number | null;
  duration_by_stage: DurationBucket[];
  duration_by_department: DurationBucket[];
  duration_by_workflow: DurationBucket[];
}

export interface CountBucket {
  key: string;
  count: number;
}

export interface RejectionBucket {
  key: string;
  decided_count: number;
  rejected_count: number;
  rejection_rate: number | null;
}

export interface BottleneckReport {
  slowest_stages: DurationBucket[];
  slowest_workflows: DurationBucket[];
  departments_causing_delay: DurationBucket[];
  approver_queue_depth: UserMetrics[];
  frequently_overdue_stages: CountBucket[];
  rejection_hotspots: RejectionBucket[];
}

export interface WorkloadReport {
  approvals_per_approver: UserMetrics[];
  requests_per_department: Record<string, number>;
  requests_per_workflow: Record<string, number>;
  completed_today: number;
  completed_this_week: number;
  completed_this_month: number;
  active_workload: number;
  completed_workload: number;
  pending_workload: number;
}

export interface TrendReport {
  request_volume: TimeSeries;
  completion_trend: TimeSeries;
  approval_trend: TimeSeries;
  rejection_trend: TimeSeries;
  average_completion_time_trend: TimeSeries;
}

export interface ExecutiveKPIs {
  average_approval_seconds: number | null;
  average_workflow_completion_seconds: number | null;
  sla_compliance_percentage: number | null;
  active_requests: number;
  completed_requests: number;
  pending_approvals: number;
  overdue_approvals: number;
  rejection_rate: number | null;
  throughput_per_day: number | null;
  workflow_efficiency_score: number | null;
}

export interface DepartmentAnalytics {
  department: string;
  throughput_per_day: number | null;
  sla_compliance_percentage: number | null;
  average_approval_seconds: number | null;
  active_workload: number;
  backlog_count: number;
}

export interface OperationalAnalyticsFilters {
  department?: string;
  request_type?: string;
  created_after?: string;
  created_before?: string;
}

File: app/api/schemas/operational_analytics.py
"""HTTP-only schemas for the ``/analytics/operational/*`` resources
(Milestone 12).

``app.analytics.operational_dto`` (``SLAMetrics``, ``PendingApprovalAge``,
``ApprovalDelayReport``, ``DurationBucket``, ``BottleneckReport``,
``CountBucket``, ``RejectionBucket``, ``WorkloadReport``, ``TrendReport``,
``ExecutiveKPIs``, ``DepartmentAnalytics``) are all plain frozen
``@dataclass``es, not Pydantic models — exactly like ``app.analytics.dto``.
The wrappers below give them a JSON-serializable shape via
``Model.model_validate(dataclass_instance)`` (``from_attributes=True``),
never re-deriving a field. ``UserMetricsOut``/``TimeSeriesOut`` are reused
directly from ``app.api.schemas.analytics`` rather than redefined, since
``BottleneckReport.approver_queue_depth``/``WorkloadReport
.approvals_per_approver`` and every trend field are already exactly that
shape.
"""

from __future__ import annotations

from uuid import UUID

from pydantic import ConfigDict

from app.api.schemas.analytics import TimeSeriesOut, UserMetricsOut
from app.models.base import EAHBaseModel, RequestTitle, UTCDatetime
from app.models.enums import UserRole

__all__ = [
    "PendingApprovalAgeOut",
    "SLAMetricsOut",
    "DurationBucketOut",
    "ApprovalDelayReportOut",
    "CountBucketOut",
    "RejectionBucketOut",
    "BottleneckReportOut",
    "WorkloadReportOut",
    "TrendReportOut",
    "ExecutiveKPIsOut",
    "DepartmentAnalyticsOut",
]


class PendingApprovalAgeOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.PendingApprovalAge``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    stage_id: UUID
    request_id: UUID
    request_title: RequestTitle
    request_type: str
    department: str | None
    stage_name: str
    stage_order: int
    assigned_to: UUID | None
    assigned_role: UserRole | None
    created_at: UTCDatetime
    age_hours: float
    sla_hours: float | None
    is_overdue: bool


class SLAMetricsOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.SLAMetrics``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    sla_hours_override: float | None
    pending_stage_count: int
    overdue_stage_count: int
    overdue_request_count: int
    average_current_stage_age_hours: float | None
    decided_stage_count: int
    sla_breaches_decided: int
    sla_compliance_percentage: float | None
    average_total_workflow_duration_seconds: float | None
    average_approval_duration_seconds: float | None


class DurationBucketOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.DurationBucket``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    key: str
    average_seconds: float | None
    count: int


class ApprovalDelayReportOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.ApprovalDelayReport``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    longest_pending: list[PendingApprovalAgeOut]
    oldest_pending_requests: list[PendingApprovalAgeOut]
    average_approval_seconds: float | None
    median_approval_seconds: float | None
    duration_by_stage: list[DurationBucketOut]
    duration_by_department: list[DurationBucketOut]
    duration_by_workflow: list[DurationBucketOut]


class CountBucketOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.CountBucket``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    key: str
    count: int


class RejectionBucketOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.RejectionBucket``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    key: str
    decided_count: int
    rejected_count: int
    rejection_rate: float | None


class BottleneckReportOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.BottleneckReport``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    slowest_stages: list[DurationBucketOut]
    slowest_workflows: list[DurationBucketOut]
    departments_causing_delay: list[DurationBucketOut]
    approver_queue_depth: list[UserMetricsOut]
    frequently_overdue_stages: list[CountBucketOut]
    rejection_hotspots: list[RejectionBucketOut]


class WorkloadReportOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.WorkloadReport``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    approvals_per_approver: list[UserMetricsOut]
    requests_per_department: dict[str, int]
    requests_per_workflow: dict[str, int]
    completed_today: int
    completed_this_week: int
    completed_this_month: int
    active_workload: int
    completed_workload: int
    pending_workload: int


class TrendReportOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.TrendReport``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    request_volume: TimeSeriesOut
    completion_trend: TimeSeriesOut
    approval_trend: TimeSeriesOut
    rejection_trend: TimeSeriesOut
    average_completion_time_trend: TimeSeriesOut


class ExecutiveKPIsOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.ExecutiveKPIs``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    average_approval_seconds: float | None
    average_workflow_completion_seconds: float | None
    sla_compliance_percentage: float | None
    active_requests: int
    completed_requests: int
    pending_approvals: int
    overdue_approvals: int
    rejection_rate: float | None
    throughput_per_day: float | None
    workflow_efficiency_score: float | None


class DepartmentAnalyticsOut(EAHBaseModel):
    """Wraps ``app.analytics.operational_dto.DepartmentAnalytics``."""

    model_config = ConfigDict(from_attributes=True, extra="forbid", frozen=True)

    department: str
    throughput_per_day: float | None
    sla_compliance_percentage: float | None
    average_approval_seconds: float | None
    active_workload: int
    backlog_count: int
